accepted/future-releases/abstract-external-fields/feature-specification.md

```diff
@@ -0,0 +1,205 @@
+the null-safety features are introduced.  It is suboptimal, but it works.
+With `dart:ffi`, domain specific rules are used to determine which fields
+are "external", and how they should be connected with the relevant external
+entities.
+
+With null safety that approach no longer works.  For a field whose type
+annotation is a non-nullable type, e.g., `int foo;`, it is an error unless
+initialization is performed by every generative constructor, or the field
+itself is modified to have an initializing expression, e.g.,
+`int foo = 0;`.  This adds yet another element to the declaration which is
+meaningless, confusing for readers of the code, and which may be
+inconvenient to write.  It may even involve nonsensical code like
+`X foo = throw 0;` (where `X` is a type variable) in the case where no
+expression is statically known to evaluate to an instance with the required
+type.
+
+To provide a migration path for these use cases, we introduce _abstract and
+external field declarations_.
+
+
+## Feature Specification
+
+This section specifies abstract and external variables.
+
+
+### Syntax
+
+The grammar is modified as follows:
+
+```
+<topLevelDefinition> ::=
+  ... |
+  // New alternative.
+  'external' ('final' <type>? | <varOrType>) <identifierList> ';'
+
+<finalVarOrType> ::= // New rule.
+  'final' <type>? |
+  <varOrType>
+
+<declaration> ::=
+  ... |
+  // New alternative.
+  'external' ('static'? <finalVarOrType> | 'covariant' <varOrType>)
+      <identifierList> |
+  // New alternative.
+  'abstract' (<finalVarOrType> | 'covariant' <varOrType>) <identifierList>
+```
+
+
+### Static Analysis
+
+The features specified in this document are syntactic sugar, that is, they
+are specified in terms of a small program transformation that eliminates
+them.  This fully determines the further static analysis (including errors
+and warnings), and the dynamic semantics.  The transformations are as
+follows:
+
+An abstract instance variable declaration _D_ is treated as an abstract setter
+declaration and/or an abstract getter declaration. The setter is included if
+and only if _D_ is non-final. The return type of the getter and the parameter
+type of the setter, if present, is the type of _D_ (*which may be declared
+explicitly, obtained by override inference, or defaulted to `dynamic`*).  The
+parameter of the setter, if present, has the modifier `covariant` if and only if
+_D_ has the modifier `covariant`. _For example:_
+
+```dart
+abstract class A {
+  // Abstract instance variables.
+  abstract int i1, i2;
+  abstract var x;
+  abstract final int fi;
+  abstract final fx;
+  abstract covariant num cn;
+  abstract covariant var cx;
+
+  // Desugared meaning of the above.
+  int get i1;
+  void set i1(int _);
+  int get i2;
+  void set i2(int _);
+  dynamic get x;
+  void set x(dynamic _);
+  int get fi;
+  dynamic get fx;
+  num get cn;
+  void cn(covariant num _);
+  dynamic get cx;
+  void set cx(covariant dynamic _);
```

Comment:
I'd rewrite it to:
```dart
   void set cx(covariant _)
```
and then let inference handle the `var`->`dynamic` conversion.
That is, copy the abstract declaration's "type", but erasing `var` (because it'll need to be removed as return type anyway).

Same everywhere, don't convert to `dynamic` as part of the rewrite. Just rewrite, and then apply inference or other type resolutions afterwards.

Reply:
The reason why I made `dynamic` explicit is that this desugaring step relies on the result of inference. For instance, if the superclass has `var x = 2;` as source code then we need to ensure that inference has been performed on the declaration in order to either make it `int x = 2;` or, for instance `Object? x = 2;`, depending on both the initializer in the declaration itself, and the declarations of `x` in the superinterfaces of the superclass.

I believe that the form that you mention will never exist, so it seems less than helpful to try to invent it, and maybe getting it wrong in some sense..